From 966beefe08be6045bfcca26079b76a7a80413080 Mon Sep 17 00:00:00 2001
From: None <none@example.com>
Date: Thu, 28 Sep 2023 17:37:38 -0400
Subject: [PATCH] Disable libstd and libtest dylib

---
 library/std/Cargo.toml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/library/std/Cargo.toml b/library/std/Cargo.toml
index add5a20..3d670d2 100644
--- a/library/std/Cargo.toml
+++ b/library/std/Cargo.toml
@@ -10,7 +10,7 @@ edition = "2021"
 autobenches = false
 
 [lib]
-crate-type = ["dylib", "rlib"]
+crate-type = ["rlib"]
 
 [dependencies]
 alloc = { path = "../alloc", public = true }
-- 
2.48.1

